### PR TITLE
UsabilityFix: Mappings should be editable just by clicking on the row

### PR DESCRIPTION
**Pull request by Sravanksk**

### CODE REVIEW

File: src/app/netsuite/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.html [hunk 1]
```diff
@@ -19,7 +22,10 @@ <h2 mat-dialog-title>
             Select a Category from Fyle
           </strong>
         </div>
-        <mat-form-field class="mappings-dialog--field" appearance="outline">
+        <mat-form-field class="mappings-dialog--field" appearance="outline" *ngIf="editMapping">
```
Comment (Shwetabhk) [inline src/app/netsuite/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.html:25]: there is no need to mention write the same element twice, the element can be disabled from component.ts file

File: src/app/netsuite/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts [hunk 2]
```diff
@@ -179,12 +185,18 @@ export class CategoryMappingsDialogComponent implements OnInit {
       getGeneralSettings
     ]).subscribe(() => {
       that.isLoading = false;
+      const defaultNetSuiteAccount = that.editMapping ? that.netsuiteAccounts.filter(nsAccObj => nsAccObj.value === that.data.rowElement.netsuite_value)[0]: '';
```
Comment (Shwetabhk) [inline src/app/netsuite/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts:188]: why is the word default used here?

File: src/app/netsuite/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts [hunk 3]
```diff
@@ -179,12 +185,18 @@ export class CategoryMappingsDialogComponent implements OnInit {
...
+      const defaultCCCAccount = that.editMapping ? that.cccAccounts.filter(cccObj => cccObj.value === that.data.rowElement.ccc_value)[0]: '';
```
Comment (Shwetabhk) [inline src/app/netsuite/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts:189]: same?

File: src/app/netsuite/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts [hunk 4]
```diff
@@ -179,12 +185,18 @@ export class CategoryMappingsDialogComponent implements OnInit {
...
-        fyleCategory: ['', Validators.compose([Validators.required, that.forbiddenSelectionValidator(that.fyleCategories)])],
-        netsuiteAccount: ['', Validators.compose([that.forbiddenSelectionValidator(that.netsuiteAccounts)])],
-        cccAccount: ['', that.showSeparateCCCField() ? that.forbiddenSelectionValidator(that.cccAccounts) : null]
+        fyleCategory: [this.editMapping ? that.data.rowElement.fyle_value : Validators.compose([Validators.required, that.forbiddenSelectionValidator(that.fyleCategories)])],
+        netsuiteAccount: [this.editMapping ? defaultNetSuiteAccount : Validators.compose([that.forbiddenSelectionValidator(that.netsuiteAccounts)])],
+        cccAccount: [defaultCCCAccount || '', that.showSeparateCCCField() ? that.forbiddenSelectionValidator(that.cccAccounts) : null]
       });
 
+      if(that.editMapping) {
+        that.form.controls.fyleCategory.disable()
```
Comment (Shwetabhk) [inline src/app/netsuite/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts:197]: If we're already disabling here, why are we also writing the same field twice in html?

Comment (Shwetabhk) [reply src/app/netsuite/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts:197]: why use *ngIf

File: src/app/netsuite/settings/category-mappings/category-mappings.component.html [hunk 5]
```diff
@@ -8,7 +8,7 @@
       <div fxLayout="row" fxLayoutAlign="start center" class="page-info--description">
         Map category items from Fyle to their counterparts on NetSuite.<a href="https://www.fylehq.com/help/en/articles/4424248-onboarding-process-to-set-up-fyle-netsuite-integration" target="_blank" rel="noopener noreferrer" class="read-more-link" fxLayout="row" fxLayoutAlign="start center">Read more<mat-icon class="read-more-link--icon">open_in_new</mat-icon></a>
       </div>
-      <button class="page-info--primary-action" mat-flat-button color="primary" type="button" (click)="open()">
+      <button class="page-info--primary-action" mat-flat-button color="primary" type="button" (click)="open('')">
```
Comment (Shwetabhk) [inline src/app/netsuite/settings/category-mappings/category-mappings.component.html:11]: I don't think you need to pass anything in open()

**Review by Shwetabhk — CHANGES_REQUESTED** [review]

File: src/app/netsuite/settings/employee-mappings/employee-mappings-dialog/employee-mappings-dialog.component.html [hunk 6]
```diff
@@ -19,7 +22,10 @@ <h2 mat-dialog-title>
             Select an Employee from Fyle
           </strong>
         </div>
-        <mat-form-field class="mappings-dialog--form-element" appearance="outline">
+        <mat-form-field class="mappings-dialog--form-element" appearance="outline" *ngIf="editMapping">
```
Comment (Shwetabhk) [inline src/app/netsuite/settings/employee-mappings/employee-mappings-dialog/employee-mappings-dialog.component.html:25]: same element twice is not needed

File: src/app/netsuite/settings/employee-mappings/employee-mappings-dialog/employee-mappings-dialog.component.ts [hunk 7]
```diff
@@ -188,19 +189,31 @@ export class EmployeeMappingsDialogComponent implements OnInit {
...
-        netsuiteVendor: ['', that.generalSettings.employee_field_mapping === 'VENDOR' ? that.forbiddenSelectionValidator(that.netsuiteVendors) : null],
-        netsuiteEmployee: ['', that.generalSettings.employee_field_mapping === 'EMPLOYEE' ? that.forbiddenSelectionValidator(that.netsuiteEmployees) : null],
+        fyleEmployee: [that.editMapping ? that.data.rowElement.fyle_value : Validators.compose([Validators.required, that.forbiddenSelectionValidator(that.fyleEmployees)])],
+        netsuiteVendor: [that.generalSettings.employee_field_mapping === 'VENDOR' && that.editMapping ? defaultVendor : that.forbiddenSelectionValidator(that.netsuiteVendors)],
+        netsuiteEmployee: [that.generalSettings.employee_field_mapping === 'EMPLOYEE' && that.editMapping ? defaultNetsuiteEmployee : that.forbiddenSelectionValidator(that.netsuiteEmployees)],
         creditCardAccount: [defaultCCCObj || '', (that.generalSettings.corporate_credit_card_expenses_object && that.generalSettings.corporate_credit_card_expenses_object !== 'BILL') ? that.forbiddenSelectionValidator(that.cccObjects) : null]
       });
+
+      if(that.editMapping) {
+        that.form.controls.fyleEmployee.disable()
```
Comment (Shwetabhk) [inline src/app/netsuite/settings/employee-mappings/employee-mappings-dialog/employee-mappings-dialog.component.ts:203]: disabling here should be enough

File: src/app/netsuite/settings/employee-mappings/employee-mappings.component.html [hunk 8]
```diff
@@ -8,7 +8,7 @@
       <div fxLayout="row" fxLayoutAlign="start center" class="page-info--description">
         Map Employee items from Fyle to their counterparts on NetSuite. <a href="https://www.fylehq.com/help/en/articles/4424248-onboarding-process-to-set-up-fyle-netsuite-integration" target="_blank" rel="noopener noreferrer" class="read-more-link" fxLayout="row" fxLayoutAlign="start center">Read more<mat-icon class="read-more-link--icon">open_in_new</mat-icon></a>
       </div>
-      <button class="page-info--primary-action" mat-flat-button color="primary" type="button" (click)="open()">
+      <button class="page-info--primary-action" mat-flat-button color="primary" type="button" (click)="open('')">
```
Comment (Shwetabhk) [inline src/app/netsuite/settings/employee-mappings/employee-mappings.component.html:11]: I don't think we need to pass anything in open()

File: src/app/netsuite/settings/generic-mappings/generic-mappings.component.html [hunk 9]
```diff
@@ -9,7 +9,7 @@
       <div fxLayout="row" fxLayoutAlign="start center" class="page-info--description">
         Map {{ getTitle(source_field).toLowerCase() }} items from Fyle to their counterparts on NetSuite.<a href="https://www.fylehq.com/help/en/articles/4424248-onboarding-process-to-set-up-fyle-netsuite-integration" target="_blank" rel="noopener noreferrer" class="read-more-link" fxLayout="row" fxLayoutAlign="start center">Read more<mat-icon class="read-more-link--icon">open_in_new</mat-icon></a>
       </div>
-      <button class="page-info--primary-action" mat-flat-button color="primary" type="button" (click)="open()">
+      <button class="page-info--primary-action" mat-flat-button color="primary" type="button" (click)="open('')">
```
Comment (Shwetabhk) [inline src/app/netsuite/settings/generic-mappings/generic-mappings.component.html:12]: open()

File: src/app/netsuite/settings/generic-mappings/generic-mappings-dialog/generic-mappings-dialog.component.ts [hunk 10]
```diff
@@ -140,15 +139,20 @@ export class GenericMappingsDialogComponent implements OnInit {
...
+      const defaultDestinationField = that.editMapping ? that.netsuiteElements.filter(destinationField => destinationField.value === that.data.rowElement.destination.value)[0] : '';
       that.form = that.formBuilder.group({
-        sourceField: ['', Validators.compose([Validators.required, that.forbiddenSelectionValidator(that.fyleAttributes)])],
-        destinationField: ['', Validators.compose([Validators.required, that.forbiddenSelectionValidator(that.netsuiteElements)])]
+        sourceField: [that.editMapping ? that.data.rowElement.source.value : Validators.compose([Validators.required, that.forbiddenSelectionValidator(that.fyleAttributes)])],
+        destinationField: [that.editMapping ? defaultDestinationField : that.forbiddenSelectionValidator(that.netsuiteElements)],
       });
 
+      if(that.editMapping) {
+        that.form.controls.sourceField.disable()
```
Comment (Shwetabhk) [inline src/app/netsuite/settings/generic-mappings/generic-mappings-dialog/generic-mappings-dialog.component.ts:153]: This should be enough, no need to *ngIf on html

File: src/app/netsuite/settings/generic-mappings/generic-mappings-dialog/generic-mappings-dialog.component.ts [hunk 11]
```diff
@@ -140,15 +139,20 @@ export class GenericMappingsDialogComponent implements OnInit {
     that.isLoading = true;
     // TODO: remove promises and do with rxjs observables
     forkJoin([
-      getFyleAttributes,
-      netsuitePromise
+      from(getFyleAttributes),
+      from(netsuitePromise)
     ]).subscribe(() => {
       that.isLoading = false;
+      const defaultDestinationField = that.editMapping ? that.netsuiteElements.filter(destinationField => destinationField.value === that.data.rowElement.destination.value)[0] : '';
```
Comment (Shwetabhk) [inline src/app/netsuite/settings/generic-mappings/generic-mappings-dialog/generic-mappings-dialog.component.ts:146]: why is this called default?

File: src/app/netsuite/settings/generic-mappings/generic-mappings-dialog/generic-mappings-dialog.component.ts [hunk 12]
```diff
@@ -140,15 +139,20 @@ export class GenericMappingsDialogComponent implements OnInit {
     that.isLoading = true;
     // TODO: remove promises and do with rxjs observables
     forkJoin([
-      getFyleAttributes,
-      netsuitePromise
+      from(getFyleAttributes),
```
Comment (Shwetabhk) [inline src/app/netsuite/settings/generic-mappings/generic-mappings-dialog/generic-mappings-dialog.component.ts:142]: why is from needed?

File: src/app/netsuite/settings/employee-mappings/employee-mappings-dialog/employee-mappings-dialog.component.ts [hunk 13]
```diff
@@ -188,19 +189,31 @@ export class EmployeeMappingsDialogComponent implements OnInit {
       from(getGeneralMappings)
     ]).subscribe((res) => {
       const defaultCCCObj = that.cccObjects.filter(cccObj => cccObj.value === that.generalMappings.default_ccc_account_name)[0];
+      const defaultVendor = that.editMapping ? that.netsuiteVendors.filter(vendor => vendor.value === that.data.rowElement.netsuite_value)[0] : '';
```
Comment (Shwetabhk) [inline src/app/netsuite/settings/employee-mappings/employee-mappings-dialog/employee-mappings-dialog.component.ts:192]: why is this called default?

File: src/app/netsuite/settings/generic-mappings/generic-mappings.component.html (hunk 9 above)
Comment (Sravanksk) [reply src/app/netsuite/settings/generic-mappings/generic-mappings.component.html:12]: not passing anything in open() is throwing an error when the app is building `Expected 1 arguments, but got 0.`

**Review by Shwetabhk — APPROVED** [review]

**Pull request merged**